contrib/golang/filters/network/source/go/pkg/network/shim.go

```diff
@@ -0,0 +1,260 @@
+ * limitations under the License.
+ */
+
+package network
+
+/*
+// ref https://github.com/golang/go/issues/25832
+
+#cgo CFLAGS: -I../api
+#cgo linux LDFLAGS: -Wl,-unresolved-symbols=ignore-all
+#cgo darwin LDFLAGS: -Wl,-undefined,dynamic_lookup
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "api.h"
+*/
+import "C"
+import (
+	"errors"
+	"runtime"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"unsafe"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/envoyproxy/envoy/contrib/golang/common/go/api"
+	"github.com/envoyproxy/envoy/contrib/golang/common/go/utils"
+)
+
+var (
+	// ref: https://golang.org/cmd/cgo/
+	// The size of any C type T is available as C.sizeof_T, as in C.sizeof_struct_stat.
+	CULLSize uintptr = C.sizeof_ulonglong
+
+	ErrDupRequestKey = errors.New("dup request key")
+
+	DownstreamFilters = &DownstreamFilterMap{}
+	UpstreamFilters   = &UpstreamFilterMap{}
+
+	configIDGenerator uint64
+	configCache       = &sync.Map{} // uint64 -> *anypb.Any
+
+	libraryID string
+)
+
+func CreateUpstreamConn(addr string, filter api.UpstreamFilter) {
+	h := uint64(uintptr(cgoAPI.UpstreamConnect(libraryID, addr)))
+
+	// NP: make sure filter will be deleted.
+	runtime.SetFinalizer(filter, func(f api.UpstreamFilter) {
+		cgoAPI.UpstreamFinalize(unsafe.Pointer(uintptr(h)), api.NormalFinalize)
+	})
+
+	// TODO: handle error
+	_ = UpstreamFilters.StoreFilter(h, filter)
+}
+
+//export envoyGoFilterOnNetworkFilterConfig
+func envoyGoFilterOnNetworkFilterConfig(libraryIDPtr uint64, libraryIDLen uint64, configPtr uint64, configLen uint64) uint64 {
+	buf := utils.BytesToSlice(configPtr, configLen)
+	var any anypb.Any
+	proto.Unmarshal(buf, &any)
+
+	libraryID = strings.Clone(utils.BytesToString(libraryIDPtr, libraryIDLen))
+	configID := atomic.AddUint64(&configIDGenerator, 1)
+	configCache.Store(configID, GetNetworkFilterConfigParser().ParseConfig(&any))
+
+	return configID
+}
+
+//export envoyGoFilterOnDownstreamConnection
+func envoyGoFilterOnDownstreamConnection(wrapper unsafe.Pointer, pluginNamePtr uint64, pluginNameLen uint64,
+	configID uint64) uint64 {
+	filterFactoryConfig, ok := configCache.Load(configID)
+	if !ok {
+		// TODO: panic
+		return uint64(api.NetworkFilterStopIteration)
+	}
+	pluginName := strings.Clone(utils.BytesToString(pluginNamePtr, pluginNameLen))
+	filterFactory := GetNetworkFilterConfigFactory(pluginName).CreateFactoryFromConfig(filterFactoryConfig)
+
+	cb := &connectionCallback{
+		wrapper:   wrapper,
+		writeFunc: cgoAPI.DownstreamWrite,
+		closeFunc: cgoAPI.DownstreamClose,
+		infoFunc:  cgoAPI.DownstreamInfo,
+	}
+	filter := filterFactory.CreateFilter(cb)
+
+	// NP: make sure filter will be deleted.
+	runtime.SetFinalizer(filter, func(ff api.DownstreamFilter) {
+		cgoAPI.DownstreamFinalize(unsafe.Pointer(uintptr(wrapper)), api.NormalFinalize)
+	})
+
+	// TODO: handle error
+	_ = DownstreamFilters.StoreFilter(uint64(uintptr(wrapper)), filter)
```

Comment:
I didn't find when `DeleteFilter` is called to clear huh?

Reply:
delete on connection close